### PR TITLE
Licenses

### PR DESCRIPTION
**Pull request by jschaul**
Adds automation to generate license headers in haskell/rust source files. See https://github.com/zinfra/backend-issues/issues/1154 for details.

A separate PR will be made that applies this change to all files if this is approved.

See also #981

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -36,11 +36,13 @@ haddock-shallow:
 # formats all Haskell files (which don't contain CPP)
 .PHONY: format
 format:
+	licensure -p
 	./tools/ormolu.sh
 
 # formats all Haskell files even if local changes are not committed to git
 .PHONY: formatf
 formatf:
+	licensure -p
```
Comment (fisx) [inline Makefile]: If we do this, we should also add a CI check that running licensure is idempotent.

File: docs/developer/dependencies.md [hunk 2]
```diff
@@ -95,6 +95,23 @@ sudo apt install rustc cargo -y
...
+
+## Formatting Haskell files
+
+You need `ormolu` on your PATH, get it with `stack install ormolu`
+
+## Generating license headers
+
+We use [`licensure`](https://github.com/chasinglogic/licensure).
+
+As of 2020-02-18, you need the latest version:
```
Comment (fisx) [inline docs/developer/dependencies.md]:
```suggestion
To get it:
```

File: docs/developer/dependencies.md [hunk 3]
```diff
@@ -95,6 +95,23 @@ sudo apt install rustc cargo -y
...
+
+```
+git clone https://github.com/chasinglogic/licensure
+cd licensure
+cargo install --path .
```
Comment (fisx) [inline docs/developer/dependencies.md]:
```suggestion
git checkout 96d7fad9201e19e5193bf7435ab5fdf4a6ad4685  # master as of 2020-02-18
cargo install --path .
```

File: Makefile (hunk 1 above)
Comment (mheinzel) [reply Makefile:45]: It actually isn't idempotent, see my review.

**Comment by jschaul** [comment]
@mheinzel have another look, I added a makefile target that ensures headers are not added multiple times. It's not as good as idempotence, of course, but perhaps "good enough" to allow someone creating a new file to add a license header without going copy-pasting.

File: Makefile [hunk 4]
```diff
@@ -48,6 +48,18 @@ formatf:
...
+add-license:
+	# For any Haskell or Rust file that doesn't mention AGPL yet,
+	#   add a license header.
+	# It's your own reponsibility to keep ormolu happy.
+	for file in $$(git grep -L "GNU Affero General Public License" | grep '\.hs$$\|\.rs$$'); do \
+		echo "Adding license to $${file}."; \
+		licensure -i $${file}; \
+	done;
+	@echo ""
+	@echo "you most probably want to run 'make formatf' now to keep ormolu happy"
```
Comment (jschaul) [inline Makefile:61]: Of course this task could directly run ormolu, but I think it's preferrable to leave people the option to check their changes before having ormolu wreak havoc (potentially). So I prefer to separate these steps out and just do an echo here.

**Pull request merged**